File: api/pricing.py
"""Single source of truth for Solar Operator per-array volume pricing.

Billing is graduated/tiered: each band of arrays is priced on the arrays that
fall *within* that band (the way AWS / Stripe / Twilio volume pricing works),
NOT a single whole-bill discount. Graduated tiers can never cross the floor, so
there is no revenue cliff.

The SAME table feeds two places that must agree to the penny:
  1. The live Stripe Price (scripts/create_stripe_prices.py builds a Stripe
     graduated tiered price from TIERS). Stripe is the source of truth for the
     actual charge — it applies these tiers automatically given the quantity.
  2. The dashboard billing estimate (api/account.billing_summary) calls
     compute_monthly_cents() so the operator's "next charge" preview matches the
     Stripe invoice exactly.

Pricing shape (Ford, Jun 9 '26): "10% off for every 50 past 50", 30% cap.
  arrays 1–50     $15.00   (0% off — full)
  arrays 51–100   $13.50   (10% off)
  arrays 101–150  $12.00   (20% off)
  arrays 151+     $10.50   (30% off — floor/cap)

To change pricing, edit TIERS here ONLY, then re-run create_stripe_prices.py to
mint a new Stripe price and update STRIPE_ARRAY_PRICE_ID. Nothing else hardcodes
the numbers.
"""
from __future__ import annotations

# Each tier: (up_to, unit_amount_cents). `up_to` is the inclusive array count at
# which this tier's unit price stops applying; the final tier uses None to mean
# "infinity" (every array beyond the prior boundary). Ordered ascending.
# This mirrors Stripe's `tiers` array with tiers_mode="graduated".
TIERS: list[tuple[int | None, int]] = [
    (50,   1500),  # arrays 1–50    @ $15.00  (0% off)
    (100,  1350),  # arrays 51–100  @ $13.50  (10% off)
    (150,  1200),  # arrays 101–150 @ $12.00  (20% off)
    (None, 1050),  # arrays 151+    @ $10.50  (30% off — cap)
]

FULL_UNIT_CENTS = TIERS[0][1]  # $15.00 — the undiscounted per-array price


def compute_monthly_cents(array_count: int) -> int:
    """Total monthly cents for `array_count` arrays under graduated tiers.

    Mirrors Stripe tiers_mode="graduated": each tier's unit price applies only to
    the arrays that fall within that tier's band. Returns 0 for a count <= 0
    (comped/zero-array accounts); the Stripe subscription enforces its own
    quantity>=1 minimum separately — this is a pure pricing function.
    """
    if array_count <= 0:
        return 0
    remaining = array_count
    prev_bound = 0
    total = 0
    for up_to, unit in TIERS:
        if up_to is None:
            total += remaining * unit
            remaining = 0
            break
        band = up_to - prev_bound
        take = min(remaining, band)
        total += take * unit
        remaining -= take
        prev_bound = up_to
        if remaining <= 0:
            break
    return total


def blended_unit_cents(array_count: int) -> int:
    """Average per-array cents (total / count), rounded. For display only."""
    if array_count <= 0:
        return FULL_UNIT_CENTS
    return round(compute_monthly_cents(array_count) / array_count)


def stripe_tiers() -> list[dict]:
    """TIERS as Stripe Price `tiers` payload (tiers_mode='graduated')."""
    out: list[dict] = []
    for up_to, unit in TIERS:
        out.append({"up_to": "inf" if up_to is None else up_to,
                    "unit_amount": unit})
    return out

File: web/onboarding/src/pricing.ts
// Volume/graduated per-array pricing — TS mirror of api/pricing.py.
// MUST stay in sync with the backend table and the live Stripe graduated price.
// Graduated: each band's unit price applies only to the arrays within that band.
//
//   arrays 1–50     $15.00  (0% off — full)
//   arrays 51–100   $13.50  (10% off)
//   arrays 101–150  $12.00  (20% off)
//   arrays 151+     $10.50  (30% off — cap)

type Tier = { upTo: number | null; unitCents: number };

export const TIERS: Tier[] = [
  { upTo: 50, unitCents: 1500 },
  { upTo: 100, unitCents: 1350 },
  { upTo: 150, unitCents: 1200 },
  { upTo: null, unitCents: 1050 },
];

export const FULL_UNIT_CENTS = TIERS[0].unitCents; // $15.00

/** Total monthly cents for a given array count under graduated tiers. */
export function monthlyCents(count: number): number {
  if (count <= 0) return 0;
  let remaining = count;
  let prevBound = 0;
  let total = 0;
  for (const { upTo, unitCents } of TIERS) {
    if (upTo === null) {
      total += remaining * unitCents;
      break;
    }
    const take = Math.min(remaining, upTo - prevBound);
    total += take * unitCents;
    remaining -= take;
    prevBound = upTo;
    if (remaining <= 0) break;
  }
  return total;
}

/** Whole-dollar monthly total (Stripe prices are whole-dollar units here). */
export function monthlyDollars(count: number): number {
  return Math.round(monthlyCents(count) / 100);
}

/** True once the count earns any volume discount (past the first tier). */
export function hasVolumeDiscount(count: number): boolean {
  return count > (TIERS[0].upTo ?? Infinity);
}

/** What the same count would cost with no volume discount (flat $15). */
export function flatMonthlyCents(count: number): number {
  return Math.max(0, count) * FULL_UNIT_CENTS;
}

/** Monthly cents saved by volume pricing vs the flat $15 rate. */
export function savingsCents(count: number): number {
  return flatMonthlyCents(count) - monthlyCents(count);
}

/** Average per-array cents actually paid (total / count), rounded. */
export function blendedUnitCents(count: number): number {
  if (count <= 0) return FULL_UNIT_CENTS;
  return Math.round(monthlyCents(count) / count);
}

/** Discount percentage off the flat rate, rounded (0 below first tier). */
export function discountPct(count: number): number {
  const flat = flatMonthlyCents(count);
  if (flat <= 0) return 0;
  return Math.round((savingsCents(count) / flat) * 100);
}

export interface BandRow {
  /** Human label for the band, e.g. "1–50" or "151+". */
  label: string;
  /** Per-array price in this band, cents. */
  unitCents: number;
  /** Percent off the full $15 rate for this band (0, 10, 20, 30). */
  offPct: number;
  /** How many of THIS operator's arrays fall in the band (0 if none yet). */
  filled: number;
  /** Total arrays the band can hold (Infinity for the final band). */
  capacity: number;
  /** filled * unitCents — what this band contributes to the bill. */
  subtotalCents: number;
  /** True if the operator's count reaches into this band at all. */
  active: boolean;
}

/**
 * Per-band breakdown of how `count` arrays fill the graduated tiers.
 * Always returns all four bands so the UI can render the full ladder and
 * light up the active ones as the count grows.
 */
export function tierBreakdown(count: number): BandRow[] {
  const n = Math.max(0, count);
  let prevBound = 0;
  const rows: BandRow[] = [];
  for (const { upTo, unitCents } of TIERS) {
    const capacity = upTo === null ? Infinity : upTo - prevBound;
    const lower = prevBound + 1;
    const label = upTo === null ? `${lower}+` : `${lower}–${upTo}`;
    const remainingAtBand = Math.max(0, n - prevBound);
    const filled = Math.min(remainingAtBand, capacity);
    const offPct = Math.round((1 - unitCents / FULL_UNIT_CENTS) * 100);
    rows.push({
      label,
      unitCents,
      offPct,
      filled,
      capacity,
      subtotalCents: filled * unitCents,
      active: filled > 0,
    });
    if (upTo !== null) prevBound = upTo;
  }
  return rows;
}
